Mangled residue data in Amber export with single-residue systems

**Description**

Systems with a single residue have that residue written out with the name `RES` no matter what is stored in the topology.

**Reproduction**

```
import parmed

from openff.toolkit import Molecule, ForceField
from openff.interchange import Interchange


def test_residue_name():
    force_field = ForceField("openff-2.0.0.offxml")
    
    molecule = Molecule.from_smiles("CCO")
    
    for atom in molecule.atoms:
        atom.metadata['residue_name'] = "FOO"
    
    molecule.add_default_hierarchy_schemes()
    
    Interchange.from_smirnoff(force_field, [molecule]).to_prmtop("test.prmtop")
    
    assert [residue.name for residue in parmed.load_file("test.prmtop").residues][0] == ['FOO']
```

```console
$ pytest test_residue.py
========================================== test session starts ==========================================
platform darwin -- Python 3.9.16, pytest-7.2.1, pluggy-1.0.0
Using --randomly-seed=2898234765
rootdir: /Users/mattthompson/software/openff-interchange, configfile: pytest.ini
plugins: xdist-3.2.0, nbval-0.9.6, randomly-3.12.0, cov-4.0.0, anyio-3.6.2
collected 1 item

test_residue.py F                                                                                 [100%]

=============================================== FAILURES ================================================
___________________________________________ test_residue_name ___________________________________________

    def test_residue_name():
        force_field = ForceField("openff-2.0.0.offxml")

        molecule = Molecule.from_smiles("CCO")

        for atom in molecule.atoms:
            atom.metadata['residue_name'] = "FOO"

        molecule.add_default_hierarchy_schemes()

        Interchange.from_smirnoff(force_field, [molecule]).to_prmtop("test.prmtop")

>       assert [residue.name for residue in parmed.load_file("test.prmtop").residues][0] == ['FOO']
E       AssertionError: assert 'RES' == ['FOO']

test_residue.py:19: AssertionError
=========================================== warnings summary ============================================
test_residue.py::test_residue_name
test_residue.py::test_residue_name
  /Users/mattthompson/mambaforge/envs/openff-interchange-env/lib/python3.9/site-packages/packaging/version.py:111: DeprecationWarning: Creating a LegacyVersion has been deprecated and will be removed in the next major release
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================== short test summary info ========================================
FAILED test_residue.py::test_residue_name - AssertionError: assert 'RES' == ['FOO']
===================================== 1 failed, 2 warnings in 6.64s =====================================
```

https://github.com/openforcefield/openff-interchange/blame/v0.2.3/openff/interchange/interop/internal/amber.py#L484-L491

--- a/openff/interchange/interop/internal/amber.py
+++ b/openff/interchange/interop/internal/amber.py
@@ -486,14 +486,10 @@ def to_prmtop(interchange: "Interchange", file_path: Union[Path, str]):
         text_blob = "".join([str(val).rjust(8) for val in nonbonded_parm_indices])
         _write_text_blob(prmtop, text_blob)
 
-        residue_names = (
-            [
-                residue.residue_name
-                for residue in interchange.topology.hierarchy_iterator("residues")
-            ]
-            if NRES > 1
-            else ["RES"]
-        )
+        residue_names = [
+            getattr(residue, "residue_name", "RES")
+            for residue in interchange.topology.hierarchy_iterator("residues")
+        ]
         prmtop.write("%FLAG RESIDUE_LABEL\n" "%FORMAT(20a4)\n")
         text_blob = "".join([val.ljust(4) for val in residue_names])
         _write_text_blob(prmtop, text_blob)

--- a/openff/interchange/tests/interoperability_tests/internal/test_amber.py
+++ b/openff/interchange/tests/interoperability_tests/internal/test_amber.py
@@ -106,13 +106,14 @@ class TestPRMTOP(_BaseTest):
 
 
 class TestAmberResidues(_BaseTest):
-    def test_single_residue_system(self, sage, ethanol):
-        """
-        Ensure a single-molecule system without specified residues writes something that ParmEd can read.
+    def test_single_residue_system_residue_name(self, tmp_path, sage, ethanol):
+        for atom in ethanol.atoms:
+            atom.metadata["residue_name"] = "YUP"
 
-        See https://github.com/openforcefield/openff-interchange/pull/538#issue-1404910624
-        """
-        Interchange.from_smirnoff(sage, [ethanol]).to_prmtop("molecule.prmtop")
+        ethanol.add_default_hierarchy_schemes()
 
-        # Use ParmEd as a soft validator of this file format
-        parmed.amber.AmberParm("molecule.prmtop")
+        Interchange.from_smirnoff(sage, [ethanol]).to_prmtop("test.prmtop")
+
+        residue_names = [r.name for r in parmed.load_file("test.prmtop").residues]
+
+        assert residue_names == ["YUP"]

